File: embedding/embedding.py
#!/usr/bin/env python3
# Embedding quality metrics for omnibenchmark.
#
# Implementation notes
# --------------------
# - All metrics require >= 2 labels; return NaN otherwise.

import argparse
import sys
from pathlib import Path
from typing import Callable
import polars as pl
import json

sys.path.insert(0, str(Path(__file__).parent / "../src"))  # vendored `common` (src/common) + module-local writers
from common import cli  # noqa: E402

def parse_args():
    # We own the parser; src/common/cli injects the shared contract (base args + the
    # `EMBED-M` stage I/O from common/schema). This module's method params are
    # hand-rolled below, so the whole CLI stays visible here.
    p = argparse.ArgumentParser(description="EMBED-M module (scanpy-backed)")
    cli.add_base_args(p)              # --output_dir, --name
    cli.add_stage_args(p, "EMBED-M")  # --pcas_tsv, --rawdata_clusters_truth
    #p.add_argument("--number_selected", type=int, required=True, help="Number of features to select")
    return p.parse_args()


from sklearn.metrics import (
    calinski_harabasz_score,
    davies_bouldin_score,
    silhouette_score,
)

# Add or remove metrics here; all must have signature fn(X, labels) -> float.
METRICS: dict[str, Callable] = {
    "silhouette": silhouette_score,
    "davies_bouldin": davies_bouldin_score,
    "calinski_harabasz": calinski_harabasz_score,
}


def main() -> None:
    args = parse_args()

    # logging
    print(f"Output directory: {args.output_dir}") 
    print(f"Module name: {args.name}")
    print(f"pcas_tsv: {args.pcas_tsv}")
    print(f"rawdata_clusters_truth: {args.rawdata_clusters_truth}")

    pca_df = pl.read_csv(args.pcas_tsv, separator="\t")
    truth_df = pl.read_csv(args.rawdata_clusters_truth, separator="\t")

    # Align embedding rows with truth labels by cell_id.
    merged = pca_df.join(
        truth_df.select(["cell_id", "truths"]), on="cell_id", how="inner"
    )

    embedding = merged.drop(["cell_id", "truths"]).to_numpy()
    aligned_labels = merged["truths"].cast(str).to_numpy()
    n_cells = len(merged)
    n_labels = merged["truths"].n_unique()
    n_dropped = len(pca_df) - n_cells

    if n_labels >= 2:
        scores = {
            name: float(fn(embedding, aligned_labels)) for name, fn in METRICS.items()
        }
    else:
        scores = {name: float("nan") for name in METRICS}

    output_dir = Path(args.output_dir)
    output_dir.mkdir(parents=True, exist_ok=True) 

    with open(args.output_dir / f"{args.name}_embedding_metrics.json", "w") as fh:
        json.dump(
            {
                "n_cells": n_cells,
                "n_labels": n_labels,
                "n_dropped": n_dropped,
                **scores,
            },
            fh,
            indent=2,
        )


if __name__ == "__main__":
    main()

File: embedding/embedding.R
#!/usr/bin/env Rscript
# Embedding quality metrics (R/poem) for omnibenchmark.
#
# Implementation notes
# --------------------
# - All metrics require >= 2 labels; returns NA otherwise (handled by poem).

suppressPackageStartupMessages({
  library(poem)
  library(jsonlite)
  library(data.table)
})


# Add or remove metrics here; must be valid for level = "dataset".
METRICS <- c(
  "meanSW",
  "meanClassSW",
  "pnSW",
  "minClassSW",
  "cdbw",
  "cohesion",
  "compactness",
  "sep",
  "dbcv"
)

# arg parsing
source("src/common/cli.R")
p <- arg_parser("EMBED-M module")
p <- add_base_args(p)                    # --output_dir, --name
p <- add_stage_args(p, "EMBED-M")     # the stage I/O contract
# your own method params — argparser directly (its add_argument requires `help`):
args <- parse_args(p)                    # argparser's own parser

# logging
cat(sprintf("Full command: %s\n", paste(commandArgs(trailingOnly = FALSE), collapse = " ")))
cat(sprintf("LOG: command line args\n----------------------------------\n"))
for (i in 1:length(args)) {
  cat(sprintf("  %s: %s\n", names(args)[i], args[[i]]))
}
cat(sprintf("----------------------------------\n"))


dir.create(args$output_dir, showWarnings = FALSE, recursive = TRUE)

pca <- fread(args$pcas_tsv, header = TRUE)
truth <- fread(args$rawdata_clusters_truth, header = TRUE)

# Align embedding rows with truth labels by cell_id.
idx <- match(pca$cell_id, truth$cell_id)
mask <- !is.na(idx)
aligned_embedding <- pca[mask, !"cell_id"]
aligned_labels <- as.factor(truth$truths[idx[mask]])

n_cells <- sum(mask)
n_labels <- length(unique(aligned_labels))
n_dropped <- sum(!mask)

if (n_labels >= 2) {
  result_df <- getEmbeddingMetrics(
    aligned_embedding,
    labels = aligned_labels,
    metrics = METRICS,
    level = "dataset"
  )
  scores <- as.list(result_df[1, ])
  scores <- lapply(scores, as.numeric)
} else {
  scores <- setNames(as.list(rep(NA_real_, length(METRICS))), METRICS)
}

result <- c(
  list(n_cells = n_cells, n_labels = n_labels, n_dropped = n_dropped),
  scores
)
out <- file.path(
  args$output_dir,
  sprintf("%s_embedding_metrics.json", args$name)
)
writeLines(toJSON(result, auto_unbox = TRUE, pretty = TRUE), out)
